src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -31,42 +29,46 @@ def create_from_stripped_path(cls, path: PurePath) -> "PythonModule":
         )
         return cls(module_name_with_slashes.as_posix().replace("/", "."))
 
-    @property
-    def name_as_path(self) -> PurePath:
-        return PurePath(self.module.replace(".", "/"))
 
-    def address_spec(self, *, source_root: str) -> AscendantAddresses:
-        """The spec for all candidate targets which could feasibly own the module.
-
-        This uses AscendantAddresses because targets can own files in subdirs (e.g. rglobs). We also
-        use the package path, e.g. `helloworld/util/__init__.py`, rather than the module path to
-        ensure that we capture all possible targets. It is okay if this directory does not actually
-        exist.
-        """
-        return AscendantAddresses(directory=str(PurePath(source_root) / self.name_as_path))
-
-    def possible_stripped_paths(self) -> Iterator[PurePath]:
-        """Given a module like `helloworld.util`, convert it back to its possible paths.
+@dataclass(frozen=True)
+class FirstPartyModuleToAddressMapping:
+    mapping: FrozenDict[str, Address]
 
-        Each module has either 2 or 4 possible paths. For example, given the module
-        `helloworld.util`:
+    def address_for_module(self, module: str) -> Optional[Address]:
+        target = self.mapping.get(module)
+        if target is not None:
+            return target
+        # If the module is not found, try the parent, if any. This is to accommodate `from`
+        # imports, where we don't care about the specific symbol, but only the module. For example,
+        # with `from typing import List`, we only care about `typing`.
+        # Unlike with third party modules, we do not look past the direct parent.
+        if "." not in module:
+            return None
+        parent_module = module.rsplit(".", maxsplit=1)[0]
+        return self.mapping.get(parent_module)
 
-        - helloworld/util.py
-        - helloworld/util/__init__.py
-        - helloworld.py
-        - helloworld/__init__.py
 
-        The last two possible paths look at the parent module, if any. This is to accommodate `from`
-        imports, where we don't care about the specific symbol, but only the module. For example,
-        with `from typing import List`, we only care about `typing`.
-        """
-        module_name_with_slashes = PurePath(self.module.replace(".", "/"))
-        yield self.name_as_path.with_suffix(".py")
-        yield self.name_as_path / "__init__.py"
-        parent = module_name_with_slashes.parent
-        if str(parent) != ".":
-            yield parent.with_suffix(".py")
-            yield parent / "__init__.py"
+@rule
+async def map_first_party_modules_to_addresses() -> FirstPartyModuleToAddressMapping:
+    all_targets = await Get[Targets](AddressSpecs([DescendantAddresses("")]))
+    candidate_targets = tuple(tgt for tgt in all_targets if tgt.has_field(PythonSources))
```

Comment:
It would be good to hydrate sources lazily if possible, as that would allow you to avoid hydrating sources for targets that couldn't possibly own something. I expect that most of the speedup here was the removal of the non-memoizable `AscendantAddresses` calls.

One way to do that might be to have your `FirstPartyModuleToAddressMapping` only hold onto _candidate_ targets, based on having applied source roots to figure out which modules they "might" contain (because the target is a parent of where a module might live).

Then `map_module_to_address` would actually hydrate sources of candidates to get the final exact match.

Reply:
> I expect that most of the speedup here was the removal of the non-memoizable AscendantAddresses calls

> It would be good to hydrate sources lazily if possible

Okay. I think that's best in a followup so that we can properly benchmark and experiment with different implementations.